src/main/java/algorithms/curated170/medium/numberofconnectedcomponents/NumberOfConnectedComponentsUnionFind.java

```diff
@@ -0,0 +1,51 @@
+package algorithms.curated170.medium.numberofconnectedcomponents;
+
+public class NumberOfConnectedComponentsUnionFind {
+
+    int roots[];
+
+    public int countComponents(int[][] edges, int n) {
+        if (n <= 1) {
+            return 1;
+        }
+
+        roots = new int[n];
+        for (int i = 0; i < n; i++) {
+            roots[i] = i;
+        }
+
+        for (int[] edge : edges) {
+            int r1 = findRoot(edge[0]);
+            int r2 = findRoot(edge[1]);
+
+            if (r1 != r2) {
+                roots[r1] = r2;
+                n--;
+            }
+        }
+
+        return n;
+    }
+
+    private int findRoot(int key) {
+        if (key == roots[key])
+        {
+            return key;
+        }
+        return findRoot(roots[key]);
+    }
```

Comment:
If we use "path compression" it will take 1ms; otherwise, it takes 4ms.
 
     ```
private int findRoot(int key) {
        if (key != roots[key])
        {
            roots[key] = roots[roots[key]];
            key = roots[key];
            return findRoot(key);
        }else{
            return key;
        }
    }
```